cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+use crate::ValidationError;
+use crate::{
+    typecheck::{PolicyCheck, Typechecker},
+    types::{EntityRecordKind, Type},
+    ValidationMode, ValidatorSchema,
+};
+
+/// Data structure storing what data is needed
+/// based on the the [`RequestType`].
+/// For each request type, the [`EntityManifest`] stores
+/// a [`RootAccessTrie`] of data to retrieve.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EntityManifest<T = ()>
+where
+    T: Clone,
+{
+    /// A map from request types to [`RootAccessTrie`]s.
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub per_action: HashMap<RequestType, RootAccessTrie<T>>,
+}
+
+/// A map of data fields to [`AccessTrie`]s.
+/// The keys to this map form the edges in the access trie,
+/// pointing to sub-tries.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+pub type Fields<T> = HashMap<SmolStr, Box<AccessTrie<T>>>;
+
+/// The root of an entity slice.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub enum EntityRoot {
+    /// Literal entity ids
+    Literal(EntityUID),
+    /// A Cedar variable
+    Var(Var),
+}
+
+impl Display for EntityRoot {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            EntityRoot::Literal(l) => write!(f, "{l}"),
+            EntityRoot::Var(v) => write!(f, "{v}"),
+        }
+    }
+}
+
+/// A [`RootAccessTrie`] is a trie describing
+/// data paths to retrieve. Each edge in the trie
+/// is either a record or entity dereference.
+///
+/// If an entity or record field does not exist in the backing store,
+/// it is safe to stop loading data at that point.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RootAccessTrie<T = ()>
+where
+    T: Clone,
+{
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    /// The data that needs to be loaded, organized by root.
+    pub trie: HashMap<EntityRoot, AccessTrie<T>>,
+}
+
+/// An entity slice- tells users a tree of data to load
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AccessTrie<T = ()> {
+    /// Child data of this entity slice.
+    /// The keys are edges in the trie pointing to sub-trie values.
+    #[serde_as(as = "Vec<(_, _)>")]
+    pub children: Fields<T>,
+    /// For entity types, this boolean may be `true`
+    /// to signal that all the parents in the entity hierarchy
+    /// are required (transitively).
+    pub parents_required: bool,
```

Comment:
For clarity, since this requires all transitive ancestors, should we rename to `ancestors_required`?

Reply:
yeah